### DISCUSSION TITLE
set atom outside of React

### DISCUSSION THREAD

**Discussion by yf-yang** [pmndrs/jotai#2226] [Q&A] [answered]
I am writing some codes with both react and vanilla javascript. In order to allow vanilla javascript set an atom's state, inspired by jotai-location, I am writing the code like this. Is there any better patterns?
``` JavaScript
let listener = undefined;
const baseAtom = atom();
baseAtom.onMount = set => {
  listener = value => set(value);
  return () => { listener = undefined; };
}
// make the actual atom read only
const anAtom = atom(get => get(baseAtom));

// usage: vanilla js dom event listener
dom.onmouseenter = () => listener('abcdefg');
```

**Comment by yf-yang** [comment]
Em, I found this one may not fit with HMR 😓

**Reply by yf-yang** [reply]
The listener get set back to `undefined` after HMR

**Comment by dai-shi** [comment, marked as answer]
Leaking `listener` outside of atoms doesn't look very nice.

```js
const baseAtom = atom();
baseAtom.onMount = set => {
  const listener = value => set(value);
  dom.addEventListener('mouseenter', listener);
  return () => dom.removeEventListener('mouseenter', listener);
}
// make the actual atom read only
const anAtom = atom(get => get(baseAtom));
```

Or, maybe use the store API.

**Reply by yf-yang** [reply]
Oh, I need to mention that dom is not accessible from this file. So actually I need to import that listener to other files to use it.

**Reply by dai-shi** [reply]
`let listener = undefined;` simply doesn't work with two stores.

**Reply by yf-yang** [reply]
Direct store access works well.
